rasa/cli/test.py

```diff
@@ -167,7 +168,16 @@ def run_nlu_test(args: argparse.Namespace) -> None:
             args.model, "model", DEFAULT_MODELS_PATH
         )
 
-        test_nlu(model_path, nlu_data, output, vars(args))
+        await test_nlu(model_path, nlu_data, output, vars(args))
+
+
+def run_nlu_test(args: argparse.Namespace) -> None:
```

Comment:
As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.

What do you think? 

That's a short and not complete example of what I mean:

```python
def run_nlu_test(args: Argparse) -> None:
   rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))

async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
    # implementation
```

Reply:
> @Imod7 I don't understand this part in your proposed changes, what do you mean?
> 
> > Making async a list of functions and then adding the corresponding await statements when they are called.

@wochinge For someone who doesn't know the issue at all, I just thought it would be good to mention that I added `async` and `await` statements in the definition and the calling of a list of functions. Like a summary of some of the changes I did instead of mentioning the name of the functions one by one and where those changes took place. They can also look at the files that were changed but I thought it is helpful for someone to get a quick idea before checking the files.

Reply:
> As discussed in the other PR, I think it would be better to pass the arguments directly to `run_nlu_test_async` to make `run_nlu_test_async` more reusable. Our command-line parser library `argparse` is the only thing which encapsulates arguments into the `arparse.Namespace` type, while other code would rather pass the arguments directly.
> 
> What do you think?
> 
> That's a short and not complete example of what I mean:
> 
> ```python
> def run_nlu_test(args: Argparse) -> None:
>    rasa.utils.common.run_in_loop(run_nlu_test_async(args.nlu, args.out, args.config, ...))
> 
> async def run_nlu_test_async(data_path: Optional[Text], output_dir: Optional[Text], config: Optional[Text], ...) -> None:
>     # implementation
> ```

@wochinge Yes! I wanted to ask you about this one because I didn't know how to change it, like what type of arguments to pass. So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)? Can I not use something like **kwargs? Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

Reply:
> So it is better to pass explicitly the arguments that I will use inside the function (data_path, output_dir, config)?

Yes 👍🏻 This has the following advantage:
* it's clear to the caller which arguments are expected
* We can type annotate each parameter which then our CI can check which in turn makes our code more robust

> Because if I pass them explicitly and at one point we add an extra option then we have to add it in the parameters also.

I'd argue that it is the responsibility of the person adding the extra option to make sure it's used correctly. If you'd use `kwargs` then we'd still have to extra the value at some point in the code. This means the person adding the extra option would have to change other code anyway. 
Or maybe from a different perspective: If all our code just passes around some dictionaries then we end up with a giant black box. When you want to understand whether / how something is at some point a provided parameters in such a codebase you'd have to understand a lot of prior code. Does that make sense?

> didn't know how to change it, like what type of arguments to pass.

To see which parameters you need, you can look at the function and see where we call `args.<some value name>`. To see which type they should have you can look at the functions which our function calls and see how these values are annotated there. Consider the following example:

```python
def foo(args: argparse.Namespace) -> None:
   my_param = args.test
   bar(my_param)
   bar2(args.output_directory)
   
def bar(value: Text, some_value: Optional[float] = None) -> None:
   pass

def bar(dir: Optional[Text] = None) -> None:
   pass
```

By looking at the implementation of `foo` you can see that there are two CLI parameters `test` and `output_directory`. After looking at the type annotation in `bar` and `bar2` you can see that they have type `Text` and type `Optional[Text]`. The explicit version of `foo` would hence look like:

```python
def foo_explicit(my_param: Text, output_directory: Optional[Text) -> None:
   bar(my_param)
   bar2(output_directory)
```

Reply:
@wochinge Changed the parameters of `run_nlu_test_async` so it's called with the specific arguments and not with `argparse.Namespace` anymore. 
Some thoughts : 
- Maybe passing `vars(args)` as a first argument (which goes to `all_args: Optional[Dict]`) is not a good idea and I should : 
   - construct the `Dict` inside the `run_nlu_test_async` function by using all the available parameters or
   - construct the `Dict` inside the `run_nlu_test`, then passing it as a Dict argument in `run_nlu_test_async`
That way I avoid any black boxes that you mention in your comments.
- Default values are set in the parameters. But every time someone decides to change the default values in general, then someone also has to change the default values in `run_nlu_test_async`. Is this a good practice? Maybe these default values are set in some global variables somewhere? 
- I added the `docstring` from my understanding of the use of each function parameter.

Reply:
> Maybe passing vars(args) as a first argument (which goes to all_args: Optional[Dict]) is not a good idea and I should 

You have a very good point there. This whole `vars` thing is in my opinion not clean at all. I'm a bit worried though that this PR is getting to big, so I'd actually keep your solution for now. If you want you can tackle this in a follow-up PR?

> Default values are set in the parameters. 

I'd omit them for now as they shouldn't be required. 

> I added the docstring from my understanding of the use of each function parameter.
 Write
 
 Sounds good. Will review them in the final review 👍🏻